### DISCUSSION TITLE
C API - value binding limitations.

### DISCUSSION THREAD

**Discussion by Giorgi** [duckdb/duckdb#14476] [Ideas]
The [Values](https://duckdb.org/docs/api/c/value) API and [Prepared Statements](https://duckdb.org/docs/api/c/prepared) API lack some functions that make the following two scenarios impossible to implement. It would be nice to have them supported in the C API:

1. Writing a list with a null value

How do I bind [1,2,3, null] to a list column? The following code works for inserting [1,2,3] to the table but I'm not sure how to specify a value for null:

```c
auto intType = duckdb_create_logical_type(DUCKDB_TYPE_INTEGER);
auto var1 = duckdb_create_int8(1);
auto var2 = duckdb_create_int8(3);
auto var3 = duckdb_create_int8(5);

std::vector<duckdb_value> values =  { var1, var2, var3 };
auto listVal = duckdb_create_list_value(intType, values.data(), values.size());
state = duckdb_bind_value(preparedStatement, 1, listVal);

state = duckdb_execute_prepared(preparedStatement, &result);
```
2. Getting nested column type information

There is `duckdb_param_type` for getting the type of a parameter but there is no way to get the logical type for a parameter. Would be great to have a function for getting that too. Similar to `duckdb_appender_column_type`

**Comment by Tishj** [comment]
Good point, likely needs a `duckdb_create_null` method

Perhaps also `duckdb_create_typed_null(DUCKDB_TYPE type)` if the type has any merit when the value is null (null is not a type)

**Reply by Giorgi** [reply]
Yes, we need `duckdb_create_null_value` or something like that. If that value works with `duckdb_bind_value` (for primitive types too) you can get rid of `duckdb_bind_null` and all other `duckdb_bind_xxx` functions.

**Comment by Giorgi** [comment]
Implemented in #14515 and #14687
